File: CS/SnapModule/Module.cs
using System;
using DevExpress.ExpressApp;
using System.ComponentModel;
using System.Collections.Generic;
using DevExpress.ExpressApp.Updating;
using SnapModule.Base;
using DevExpress.ExpressApp.Design;
using SnapModule.Controllers;
using DevExpress.ExpressApp.Utils;

namespace SnapModule {
    public sealed partial class SnapModule : ModuleBase {
        public SnapModule() {
            InitializeComponent();
        }
        public override IEnumerable<ModuleUpdater> GetModuleUpdaters(IObjectSpace objectSpace, Version versionFromDB) {
            ModuleUpdater updater = new DatabaseUpdate.Updater(objectSpace, versionFromDB);
            return new ModuleUpdater[] { updater };
        }
        public override void Setup(XafApplication application) {
            base.Setup(application);
        }
        private Type _DocumentDataType;
        [TypeConverter(typeof(BusinessClassTypeConverter<IDocumentData>))]
        public Type DocumentDataType {
            get {
                return _DocumentDataType;
            }
            set {
                if (value != null) {
                    Guard.TypeArgumentIs(typeof(IDocumentData), value, "value");
                }
                _DocumentDataType = value;
            }
        }
        protected override IEnumerable<Type> GetDeclaredControllerTypes() {
            return new Type[] { 
                typeof(DocumentDataController),
                typeof(InplaceDocumentsController)
            };
        }
    }
}

File: VB/SnapModule/Module.vb
Imports System
Imports DevExpress.ExpressApp
Imports System.ComponentModel
Imports System.Collections.Generic
Imports DevExpress.ExpressApp.Updating
Imports SnapModule.Base
Imports DevExpress.ExpressApp.Design
Imports SnapModule.Controllers
Imports DevExpress.ExpressApp.Utils

Namespace SnapModule
    Public NotInheritable Partial Class SnapModule
        Inherits ModuleBase

        Public Sub New()
            InitializeComponent()
        End Sub
        Public Overrides Function GetModuleUpdaters(ByVal objectSpace As IObjectSpace, ByVal versionFromDB As Version) As IEnumerable(Of ModuleUpdater)
            Dim updater As ModuleUpdater = New DatabaseUpdate.Updater(objectSpace, versionFromDB)
            Return New ModuleUpdater() { updater }
        End Function
        Public Overrides Sub Setup(ByVal application As XafApplication)
            MyBase.Setup(application)
        End Sub
        Private _DocumentDataType As Type
        <TypeConverter(GetType(BusinessClassTypeConverter(Of IDocumentData)))> _
        Public Property DocumentDataType() As Type
            Get
                Return _DocumentDataType
            End Get
            Set(ByVal value As Type)
                If value IsNot Nothing Then
                    Guard.TypeArgumentIs(GetType(IDocumentData), value, "value")
                End If
                _DocumentDataType = value
            End Set
        End Property
        Protected Overrides Function GetDeclaredControllerTypes() As IEnumerable(Of Type)
            Return New Type() { GetType(DocumentDataController), GetType(InplaceDocumentsController) }
        End Function
    End Class
End Namespace
